dask/distributed/worker.py

```diff
@@ -427,16 +432,27 @@ def collect(self, locations):
         qkey = str(uuid.uuid1())
         queue = Queue()
         self.queues[qkey] = queue
+        locations_copy = locations.copy()  # don't mutate global locations
 
         # Send out requests for data
-        log(self.address, 'Collect data from peers', locations)
+        log(self.address, 'Collect data from peers', locations_copy)
         start = time()
         counter = 0
         with logerrors():
-            for key, locs in locations.items():
+            for key, locs in list(locations_copy.items()):
                 if key in self.data:  # already have this locally
+                    locations_copy.pop(key)
                     continue
                 worker = random.choice(tuple(locs))  # randomly select one peer
+
+                # track keys and where they are comming from
+                if worker not in self.queues_by_worker:
+                    self.queues_by_worker.update({worker: {qkey: [key]}})
+                elif qkey not in self.queues_by_worker[worker]:
+                    self.queues_by_worker[worker].update({qkey: [key]})
+                elif key not in self.queues_by_worker[worker][qkey]:
+                    self.queues_by_worker[worker][qkey].append(key)
```

Comment:
Ha, I actually did try to make a `defaultdict(defaultdict(list))` thing. My invocations didn't work but maybe this will.